Docs 5776 wsdl and raml strict validation pa (#139)
mulesoft/docs-api-manager: modules/ROOT/pages/api-proxy-landing-page.adoc

You can specify the version of the proxy you want to use by clicking the *Advanced Options* drop-down menu and selecting the proxy version that best covers your needs.

> == API Proxy and Validations

Reviewer: Please consider including:
- That the proxy validations are available for mule 3 and 4.
- That the wsdl and raml proxies have this new feature (Seems that these docs only talk about the raml/rest one)
- Which are the defaults of the validation types, depending on the proxy type (wsdl/raml/rest) and the mule version (3x/4)
- Consider saying that, for mule 3 runtimes, this feature is supported for 3.8.x and onwards.
- By enabling the validation:
in the raml proxy: the payload (against a schema), query parameters, uri parameters, headers and form parameters  are validated (There may be an apikit document talking about these validations, consider linking it if it exists).
in the wsdl proxy, the envelope is validated against the structure provided in the wsdl file.
- It should be clear that raml/rest and wsdl proxies will have the validation flags feature, but strict validation is only for mule 4 raml/rest proxy

Reviewer 2: There is no mention of Mule 3. I agree on Fede's comment. Can we add it somewhere?

For the _Which are the defaults of the validation types_, should this be linking to an API Kit doc where they have this information? We are inheriting this configuration from them

Author: Done! Thanks!

> With Mule 4, you can now perform strict validations directly from API Manager. In previous versions, you had to download the proxy from Studio, modify the validation settings, upddate the XML configuration files, and finally repackage and deploy the proxy using one of the runtimes: CloudHub, Hybrid, or Runtime Manager. This was a complicated method, and required Mule Runtime and Studio expertise.

Reviewer 2: Does it worth telling how it was done before.
It could be done directly by saying that there is a simplified method of applying validations directly from the Managed API, with no Mule/Studio experience.

Author: This is usually good information for release notes. Do we want to add it there. This is good information though.

> With Mule 4, you can now perform strict validations directly from API Manager. In previous versions, you had to download the proxy from Studio, modify the validation settings, upddate the XML configuration files, and finally repackage and deploy the proxy using one of the runtimes: CloudHub, Hybrid, or Runtime Manager. This was a complicated method, and required Mule Runtime and Studio expertise.

Reviewer 3: ```suggestion
With Mule 4, you can perform strict validations directly from API Manager rather than working through complex processes using Anypoint Studio and a runtime engine.
```

Reviewer 3: What is a "strict" validation? Can we omit that word and retain meaning?

Reviewer 3: I agree with the reviewer who commented about documenting what used to be. I've suggested a rewording to minimize that and still give context to the improvement offered by Mule 4.

> You can now modify the validation settings in API Manager when you create the proxy, and send it to any runtime to deploy it. When creating or modifying a proxy, you must use a proxy version later than 2.1.0. In Mule 4, strict validation is enabled by default.

Reviewer: Correct this line "In Mule 4, strict validation is enabled by default." it should say "In Mule 4, validations are enabled, but strict validations are disabled by default.

Author: Done! Thanks!

> You can now modify the validation settings in API Manager when you create the proxy, and send it to any runtime to deploy it. When creating or modifying a proxy, you must use a proxy version later than 2.1.0. In Mule 4, strict validation is enabled by default.

Reviewer 3: ```suggestion
You can now modify validation settings in API Manager when you create the proxy, and then deploy the proxy to any runtime. The proxy version must be later than 2.1.0, and strict validation is enabled by default.
```

> When you enable strict validations, the application compares a request with the proxy specifications. You can perform strict validations for only query parameters, only headers, or for both. For RAML and WSDL proxies, you must have already created a specification.

Reviewer: Please consider clarifying that when stict validations are on, ONLY the headers or query params specified in the api definition are the only ones allowed. On the other hand, if you enable validations (not in strict mode), the params will be validated against the api definition too, but if the proxy receives a query param or header that is not specified, that parameter is sent to the backend with the rest of them.

Author: Done! Thanks!

> When you enable strict validations, the application compares a request with the proxy specifications. You can perform strict validations for only query parameters, only headers, or for both. For RAML and WSDL proxies, you must have already created a specification.

Reviewer 3: I don't understand why we begin this para with "when you enable...," when the previous para says they are enabled by default.

Reviewer 3: It seems like this explanation of a "strict" validation should come when we first refer to that.

You can specify the version of the proxy you want to use by clicking the *Advanced Options* drop-down menu and selecting the proxy version that best covers your needs.

== API Proxy Validations

> With Mule 4, you can perform strict validations directly from API Manager. Strict validations enable the application to compare a request with proxy specifications and allow only those headers and query parameters that are specified in the API definition. 

Reviewer: This should not be the intro of the whole feature. It only talks about mule 4 strict headers. I would add an intro talking about the feature in general, saying all the runtimes that support, and the types of proxies that are supported (raml/rest and wsdl).
Also, IMHO, I would then include the default values, and finally I would include this whole text, which only talks about the specific scenarios when using mule 4 and strict validations.

You can perform strict validations for only query parameters, only headers, or for both. This feature is currently available only for RAML and WSDL proxies in Mule 4.

If you do not enable strict validations, the parameters are still validated against the API definition. However, if the proxy receives a query parameter or a header that is not specified in the definition, that parameter is irrespectively sent to the backend service.

You can  modify validation settings in API Manager when you create a proxy and deploy it using Runtime. You must use a proxy version later than 2.1.0 when modifying or creating a proxy. 

=== API Specifications for Strict Validations

> When enabling strict validations for Mule 4 Applications, you must consider its architecture before designing the API specification. If your Mule application implements a proxy or a load balancer in front, the web server might modify the client request to include additional details, such as headers. In such cases, the Mule Application denies the request due to a non-valid header error. 

Reviewer: I would change "you must consider its architecture"  with "you must consider the architecture you are including it" or something similar. We try to warn the users to consider the layers, such as proxies or load balancers around the mule runtime / or application, before enabling strict headers.

Author: Done! Thanks!

You can specify the version of the proxy you want to use by clicking the *Advanced Options* drop-down menu and selecting the proxy version that best covers your needs.

== API Proxy Validations

> You can configure API proxies to perform validations on all incoming requests against the API definition you design. Depending on the Mule runtime and API Proxy version you installed, you can modify these validation settings in API Manager and redeploy the proxy in Mule Rumtime. 

Reviewer: ```suggestion
You can configure RAML/REST/WSDL proxies to perform validations on all incoming requests against the API definition you design. Depending on the Mule runtime and API Proxy version you installed, you can modify these validation settings in API Manager and redeploy the proxy in Mule Rumtime. 
```

Author: Done! Thanks!

> You can configure API proxies to perform validations on all incoming requests against the API definition you design. Depending on the Mule runtime and API Proxy version you installed, you can modify these validation settings in API Manager and redeploy the proxy in Mule Rumtime. 

Reviewer 2: Typo here. `Mule Rumtime`

Author: Wouldn't it be nice if it were "rumtime". :-) Good catch! Thanks!

> With Mule 4, you can apply strict validations as well as regular validations. for your API proxies. Additionally, you can also edit proxies directly from API Manager based on the API proxy version. For more information, see the "Default Setting by API Proxy Versions" table in this section.

Reviewer 2: Can we link it to that header?
"Default Setting by API Proxy Versions"

Author: Ok, Done! Thanks!

=== Regular Validations

> Regular validations enable applications to compare a request with API specifications. If the proxy receives a query parameter or a header that is not specified in the definition, that parameter is irrespectively sent to the backend service. You can perform regular validations with WSDL and REST/RAML proxies. As part of the regular validations for REST/RAML proxies, the payload (schema), query parameters, URI parameters, headers, and form parameters are validated. For WSDL proxies, the envelope is validated against the structure provided in the WSDL file.

Reviewer: ```suggestion
Regular validations enable applications to compare a request with API specifications.  As part of the regular validations for REST/RAML proxies, the payload (schema), query parameters, URI parameters, headers, and form parameters are validated. For WSDL proxies, the envelope is validated against the structure provided in the WSDL file. 
If the proxy receives a parameter that is not specified in the definition, that parameter is irrespectively sent to the backend service.
```

Author: Done! Thanks!

=== Strict Validations

Strict validations enable applications to compare a request with proxy specifications and allow only those headers and query parameters that are specified in the API definition. You can perform strict validations for only query parameters, only headers, or for both. This feature is currently available only for RAML and WSDL proxies in Mule 4. 

The following tables describe the default settings for API proxies based on different Mule versions:

==== Default Setting by API Proxy Versions for Mule 3.x Applications

[%header%autowidth.spread]
|===
| Mule Version | API Proxy | Default Setting
| 3.7x and below | RAML | Disabled 
| 3.7x and below | WSDL | Disabled 
| 3.8x and above | RAML | Disabled 

> | 3.8x and above | WSDL | Disabled 

Reviewer: | Mule Version | Proxy Type | Validation Default Setting | UI Support
| 3.7.x | RAML / WSDL | Disabled | No
| 3.8.x | RAML / WSDL | Disabled | Yes
| 3.9.x | RAML / WSDL | Disabled | Yes

|===

Mule 3.x applications and API proxy compatibility synopsis:

* Only regular proxy validations are available for Mule 3.x applications.
* Validations are performed for only WSDL and RAML proxies.
* Proxies are not versioned in Mule 3.x applications.
* You can edit proxies from API Manager for Mule 3.8.x and above.

==== Default Setting by API Proxy Versions for Mule 4.x Applications

[%header%autowidth.spread]
|===

> | Mule Version | API Proxy | Default Setting

Reviewer: | Mule Version | Proxy Type | Validation Default Setting | UI Support
| 4.x| RAML 1.1.x                    | Enabled | No
| 4.x| RAML 1.2.x                   | Enabled | Yes
| 4.x| REST 2.0.x                    | Enabled | No
| 4.x| REST 2.1.0 or greater  | Enabled | Yes 
| 4.x| WSDL 1.2.0 or earlier  | Disabled | No
| 4.x| WSDL 1.2.1 or greater | Disabled | Yes

Author: Done! Thanks!

| 4.x| RAML 1.1.3 and below  | Enabled 
| 4.x| RAML 1.2.0            | Enabled 
| 4.x| REST 2.0.0            | Enabled 
| 4.x| REST 2.1.0 and above  | Enabled 
| 4.x| WSDL 1.2.0 and below  | Disabled 
| 4.x| WSDL 1.2.1 and above  | Disabled 
|===

Mule 4.x applications and API proxy compatibility synopsis:

* For Mule 4.x applications, both regular and strict validations are available.
* Validations are performed for only WSDL and RAML/REST proxies.
* Proxies are versioned in Mule 4.x applications. 
* You can edit RAML/REST proxies from API Manager for RAML proxy version 1.2.0 and REST proxy version 2.1.0 and above.

=== API Specifications for Strict Validations

When enabling strict validations for Mule 4 applications, you must consider your API architecture. If your Mule application has a proxy or a load balancer in front of it, the web server might modify the client request to include additional headers. In such cases, the Mule application denies the incoming request due to a non-valid header error. 

> 

Reviewer: ```suggestion
Take into account that CloudHub and RuntimeFabric are nowadays including headers when sending requests to the API as well.
```

Author: Done! Thanks!

==== Runtime Fabric headers

The following headers are added by Edge in Runtime Fabric:

> `headers:

Reviewer 2: Can we format this properly? Not sure if the docs page support it.

```
 x-correlation-id:
     type: string
   x-forwarded-for:
     type: string
   via:
     type: string
   x-forwarded-host:
     type: string
```

Author: That is how it is in the file, but when the content is rendered, it creates an issue.
